### DISCUSSION TITLE
Getting backtrace to work

### DISCUSSION THREAD

**Discussion by HeisenBuerger42** [vfsfitvnm/frida-il2cpp-bridge#554] [Q&A] [answered]
I lately experimented a lot with frida-il2cpp-bridge, and I think it is a very helpful tool. On my latest project I try to hook onto some functions of an android app that is running inside an emulator. In order to achieve my desired results I want to better understand the overall function calls the app is making. So I got my function, but I would like to see where it is called from. I saw that there is a backtrace functionality inside frida-il2cpp-bridge, but I wasn't really able to get it to work. I used the following code.

```javascript
Il2Cpp.backtrace().classes(BattlePreviewPopup).and().attach();
```

As far as I understand it the backtrace functionality should output the current stack when a function inside the ```BattlePreviewPopup``` class is called. But sadly it only outputs: 

![backtrace](https://github.com/user-attachments/assets/dacf4f01-8a0f-414f-b9f4-d405f6e3f1d3)

I don't really know if it is a game-specific issue since I saw that the backtrace functionality is still titled "very experimental". I would be interested in hearing if anyone had any success in using this feature. Thanks for your time.

**Comment by vfsfitvnm** [comment, marked as answer]
Hi :smile: 
The backtracer entirely relies on Frida's `Thread::backtrace` - here's how I use it: https://github.com/vfsfitvnm/frida-il2cpp-bridge/blob/a28fa2eb5f10ef4a5c0635de1d9a11ec73a7231a/src/tracer.ts#L345-L367

Unfortunately, it's not possible to go faaar back, and you also should consider that a method might not be called by another method within the IL2CPP module: that's probably why you see an empty log

**Reply by HeisenBuerger42** [reply]
Thank you for your quick reply. Really appreciate your work with this tool, ;) Your tip that the method doesn't get called within the IL2CPP module might actually be the answer to the empty log. Earlier that day I tried decompiling the lib2cpp.so with the help of il2cppdumper and ghidra. After looking at the function call graph, I saw that the function never gets called inside the .so file. Although that, might also be because I messed something up during the decompiling process. I am still pretty new to all of that stuff :D

**Comment by muvellidulMA** [comment]
Eski bir konu ama belki birilerinin ihtiyacı olabilir. Bu fonksiyon Il2Cpp modülü içerisinde bu şekilde tanımlanmış:
```
 function backtrace(mode?: Backtracer): Il2Cpp.Tracer.Configure;
 ```
Burada backtrace fonksiyonu parametre olarak "Backtracer" türünde bir değişken bekliyor. Aşağıda bu türe ait deklerasyon bulunmaktadır.
```
// tslint:disable-next-line:no-unnecessary-class
declare class Backtracer {
    /**
     * The accurate kind of backtracers rely on debugger-friendly binaries or
     * presence of debug information to do a good job, but avoid false
     * positives.
     */
    static ACCURATE: Backtracer;

    /**
     * The fuzzy backtracers perform forensics on the stack in order to guess
     * the return addresses, which means you will get false positives, but it
     * will work on any binary.
     */
    static FUZZY: Backtracer;
}
```

Sizin örneğinizde Backtracer.FUZZY veya Backtracer.ACCURATE kullanmanız gerekiyor. Sanırım bu durumda en doğru yaklaşım bir fonksiyon tanımlamak ve onu daha sonra çağırmak:
```
function safeBacktrace(mt: Il2Cpp.Method) {
    try {
        return Il2Cpp.backtrace(Backtracer.ACCURATE).methods(mt).and().attach();
    } catch (e) {
        return Il2Cpp.backtrace(Backtracer.FUZZY).methods(mt).and().attach();
    }
}
```
